### DISCUSSION TITLE
How to ensure that environment events are on the road network?

### DISCUSSION THREAD

**Discussion by phil-hei** [eclipse/mosaic#445] [Q&A]
Hi, 

I just wonder if there is an (easy) way to make sure that environment events, such as an obstacle, are placed on the road network so that vehicle can pass them by. Example: I have a defined simulation area where I want to generate random obstacles on the road. I'm able to generate this events within the simulation area, but haven't found a suitable way to ensure that they are on the road network.

**Comment by schwepmo** [comment]
Hey @phil-hei, could you extend a little bit on your question, also example configuration always help.
I assume you're using SUMO as a traffic simulator and the Environment Simulator to place your "obstacles". If that's the case then what you're trying to achieve might not be as straight forward as you're planning.
The Environment Simulator doesn't directly couple to SUMO, i.e., environment events do not get passed to the traffic simulation and thereby the road network directly. Instead, the applications running on your vehicles may react on the events (see the [Weather Warning Apps](https://github.com/eclipse/mosaic/tree/main/app/tutorials/weather-warning/src/main/java/org/eclipse/mosaic/app/tutorial)).

Alternatively, there is the option to "close" lanes using TMC's `changeLaneState` functionality (see [here](https://github.com/eclipse/mosaic/blob/66504369a967e02d3e299c82a3702168dc698582/fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/app/api/os/TrafficManagementCenterOperatingSystem.java#L67)). However, vehicles may behave weirdly if they are on a lane while it is being closed.

Lastly, if you just want a static obstacle throughout the simulation, there is the option of placing a halting vehicle (see for example [here](https://sumo-user.narkive.com/XiC6tPwD/simulation-of-obstacles-in-sumo)), which will be evaded by other vehicles automatically.

I hope these ideas help you to solve your issue, please feel free to elaborate!

**Reply by phil-hei** [reply]
Thanks for your extensive answer!
Indeed, I want to use SUMO and extend the traffic simulation with environmental events, such as obstacles on the way. Vehicles should recognize these obstacles during driving and share this information with the cloud so that other vehicles can be warned. Therefore, I just thought about how to ensure that static obstacles are placed onto the road network.
I didn't consider that these simulators are not coupled, but probably I will try out to equip vehicles with an application similar to the Weather Warning App. In this way, I can achieve what I want.
Now I just have to find a way to place static obstacles on the road network and not besides.

**Comment by kschrab** [comment]
I found a thread in the SUMO mailing list which suggests to just add a vehicle and stop it immediately at their spawn position to mimic an obstacle that other vehicles can react to and overtake it. You can either do that the "right" way, by adding a specific application which stops a vehicle as soon as it spawns, or by "hacking" the SumoAmbassador.

By spawning a vehicle and using a custom application which stops the vehicle (might not be accurate enough):

```
public void onVehicleUpdated(VehicleData previous, VehicleData updated) {
  if (previous == null)  {
    getOs().stopNow(VehicleStopMode.STOP, Long.MAX_VALUE);
  }
}
```

or by spawning a specific vehicle of type "Obstacle" and then adjusting the SumoAmbassador in line 302:

```diff
- bridge.getSimulationControl().addVehicle(vehicleId, routeId, vehicleType, laneId, departPos, departSpeed);
+ if (vehicleType.equals("Obstacle")) {
+    bridge.getSimulationControl().addVehicle(vehicleId, routeId, vehicleType, laneId, departPos, "0");
+    bridge.getVehicleControl().stop(vehicleId, routes.get(routeId).getConnectionIds().iterator().next(), vehicleRegistration.getDeparture().getDeparturePos(), vehicleRegistration.getDeparture().getDepartureLane(), Long.MAX_VALUE, VehicleStopMode.STOP);
+} else {
+    bridge.getSimulationControl().addVehicle(vehicleId, routeId, vehicleType, laneId, departPos, departSpeed);
+}
```

**Comment by phil-hei** [comment]
Just an update:
I created a component that extracts geo-positions from the according OSM road network, generates different kinds of events and store them in the environment_config.json as follows:

```
{
  "events": [
    {
      "type": {
        "sensorType": "POTHOLE",
        "value": 4.0
      },
      "location": {
        "area": {
          "type": "Rectangle",
          "a": {
            "latitude": 51.51688502792793,
            "longitude": 7.422082681087537
          },
          "b": {
            "latitude": 51.51711407207207,
            "longitude": 7.422341918912463
          }
        }
      },
      "time": {
        "start": "10 s",
        "end": "65 s"
      }
    },
    {
      "type": {
        "sensorType": "POTHOLE",
        "value": 1.0
      },
      "location": {
        "area": {
          "type": "Rectangle",
          "a": {
            "latitude": 51.51473292792793,
            "longitude": 7.424551586663976
          },
          "b": {
            "latitude": 51.51487987207207,
            "longitude": 7.424924113336023
          }
        }
      },
      "time": {
        "start": "0 s",
        "end": "87 s"
      }
    },
...
```

The position are more or less on the road, but at least close by and for the first this is sufficient.

In this direction I wonder if the API for the in-vehicle application has changed as the [Weather Warning Apps](https://github.com/eclipse/mosaic/tree/main/app/tutorials/weather-warning/src/main/java/org/eclipse/mosaic/app/tutorial) example is not working in 25.0 anymore? For example, `getOs().getBasicSensorModule().enable();` is not working anymore. Has this changed now to `getOs().getPerceptionModule().enable(new PerceptionModuleConfiguration())` or how do vehicles detect this events?

**Reply by schwepmo** [reply]
Yeah, in fact the API revolving around environment sensors has been changed, as mentioned in the changelog (https://github.com/eclipse-mosaic/mosaic/releases/tag/25.0):
*"[A+] Overhaul of API to access sensor information, such as environmental events or LiDAR point clouds."*

What do you mean by "Weather Warning Apps" example is not working anymore? As far as I can see from results of the unit and integration tests environment events are still being triggered and also sensed. Is there maybe a version mismatch with you apps, scenarios, and/or Mosaic?

**Reply by phil-hei** [reply]
Hi, thanks for the clarification. I just meant that the example don't work anymore for version 25.0 due tot he API changes.

**Reply by hoelger** [reply]
How do you notice that the example doesn't work? For me everything works as expected, when executing the Barnim Scenario.

**Reply by kschrab** [reply]
@hoelger What I found is, that the example snippets in [1] have not been corrected. I will do that asap.

[1] https://eclipse.dev/mosaic/tutorials/barnim_develop_applications/

**Reply by hoelger** [reply]
I see! :)

**Comment by phil-hei** [comment]
I have a follow-up question regarding the different sensor types provided. What is the actual purpose? For example, if I have `basicSensorModule.getStrengthOf(SensorType.RAIN)`, the sensor would only react to events that are exactly called accordingly? 
```
"type": {
        "sensorType": "RAIN",
        "value": 6
      },
```

Or is there a general sensor that react to all kind of events no matter what they are called as soon as the vehicle pass by?

**Reply by hoelger** [reply]
Unfortunately no, you have to iterate over all sensors.

see https://github.com/eclipse-mosaic/mosaic/blob/main/app/tutorials/weather-warning/src/main/java/org/eclipse/mosaic/app/tutorial/WeatherWarningApp.java#L150
